org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/ReconcileWorkingCopyOperation.java

```diff
@@ -175,41 +178,65 @@ public org.eclipse.jdt.core.dom.CompilationUnit makeConsistent(CompilationUnit w
 		if (this.ast != null)
 			return this.ast; // no need to recompute AST if known already
 
-		CompilationUnitDeclaration unit = null;
 		try {
 			JavaModelManager.getJavaModelManager().abortOnMissingSource.set(Boolean.TRUE);
 			CompilationUnit source = workingCopy.cloneCachingContents();
 			// find problems if needed
 			if (JavaProject.hasJavaNature(workingCopy.getJavaProject().getProject())
 					&& (this.reconcileFlags & ICompilationUnit.FORCE_PROBLEM_DETECTION) != 0) {
 				this.resolveBindings = this.requestorIsActive;
-				if (this.problems == null)
+				if (this.problems == null) {
 					this.problems = new HashMap<>();
-				unit =
-					CompilationUnitProblemFinder.process(
-						source,
-						this.workingCopyOwner,
-						this.problems,
-						this.astLevel != ICompilationUnit.NO_AST/*creating AST if level is not NO_AST */,
-						this.reconcileFlags,
-						this.progressMonitor);
-				if (this.progressMonitor != null) this.progressMonitor.worked(1);
-			}
-
-			// create AST if needed
-			if (this.astLevel != ICompilationUnit.NO_AST
-					&& unit !=null/*unit is null if working copy is consistent && (problem detection not forced || non-Java project) -> don't create AST as per API*/) {
+				}
 				Map<String, String> options = workingCopy.getJavaProject().getOptions(true);
-				// convert AST
-				this.ast =
-					AST.convertCompilationUnit(
-						this.astLevel,
-						unit,
-						options,
-						this.resolveBindings,
-						source,
-						this.reconcileFlags,
-						this.progressMonitor);
+				if (CompilationUnit.DOM_BASED_OPERATIONS) {
```

Comment:
Instead of a lot `if (CompilationUnit.DOM_BASED_OPERATIONS)` ii think it would be better to have different implementation Classes. However i am not familiar enough with these classes to give a substantial review on this code.

Reply:
I will try for this particular class, but it's not possible for CompilationUnit.

Reply:
That it is not as easy is not enough reason to pollute the code base with special cases for an experiment